setup.py

```diff
@@ -0,0 +1,44 @@
+# Copyright 2024 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import atexit
+import os
+import shutil
+import tempfile
+
+from setuptools import setup  # type: ignore[import]  # The Python 2.7 type check can't see this.
+
+_BUILD_DIR = None
+
+
+def ensure_unique_build_dir():
+    # type: () -> str
+    global _BUILD_DIR
+    if _BUILD_DIR is None:
+        _build_dir = tempfile.mkdtemp(prefix="pex-dist-build.")
+        atexit.register(shutil.rmtree, _build_dir, ignore_errors=True)
+        _BUILD_DIR = _build_dir
+    return _BUILD_DIR
+
+
+def unique_build_dir(name):
+    # type: (str) -> str
+    path = os.path.join(ensure_unique_build_dir(), name)
+    os.mkdir(path)
+    return path
+
+
+if __name__ == "__main__":
+    setup(
+        # The `egg_info --egg-base`, `build --build-base` and `bdist_wheel --bdist-dir` setup.py
+        # sub-command options we pass below work around the otherwise default `<CWD>/build/`
+        # directory for all three which defeats concurrency in tests.
+        options={
+            "egg_info": {"egg_base": unique_build_dir("egg_base")},
+            "build": {"build_base": unique_build_dir("build_base")},
+            "bdist_wheel": {"bdist_dir": unique_build_dir("bdist_dir")},
+        },
+        # This supports expanding the supported Python range via the _PEX_REQUIRES_PYTHON env var
+        # for testing unreleased Pythons.
+        python_requires=os.environ.get("_PEX_REQUIRES_PYTHON"),
```

Comment:
This is what necessitated the build system change from `flit` to `setuptools`. AFAICT setuptools provides the only pep-517 build backend where Requires-Dist metadata can be easily altered which is what I needed to be able to do to keep Python 3.13 support officially turned off while testing against Python 3.13 alpha, beta and rc releases from now through the end of September as the 3.13 release works towards final.

Reply:
And, although this was thoroughly vetted by tests - which caught issues with the new packaging relentlessly, the eyeball looks like so:
```
$ tox -epackage -- --additional-format sdist --additional-format wheel
ROOT: will run in automatically provisioned tox, host /home/jsirois/bin/tox.venv/bin/python3.11 is missing [requires (has)]: tox<4 (4.11.4), virtualenv<20.16 (20.25.0)
ROOT: install_deps> python -I -m pip install 'tox<4' 'tox>=3.25.1' 'virtualenv<20.16'
ROOT: provision> .tox/.tox/bin/python -m tox -epackage -- --additional-format sdist --additional-format wheel
package create: /home/jsirois/dev/pantsbuild/jsirois-pex/.tox/package
package installdeps: build
package installed: build==1.0.3,packaging==23.2,pyproject_hooks==1.0.0
package run-test-pre: PYTHONHASHSEED='2295051996'
package run-test: commands[0] | python scripts/package.py --additional-format sdist --additional-format wheel
Building Pex PEX to `dist/pex` ...
Built Pex PEX @ v2.1.156-1-g6a6e403a:
sha256: 8eeaef638aa9db10766f7805c231687b36a27dff84031a0721c336cdb451b872
  size: 3668094
Building additional distribution formats to `dist`: 1.) sdist, 2.) wheel ...
Built:
  dist/pex-2.1.156-py2.py3-none-any.whl
  dist/pex-2.1.156.tar.gz
____________________________________________________________________________________________________ summary _____________________________________________________________________________________________________
  package: commands succeeded
  congratulations :)
```

+ sdist:
    ```diff
    $ diff -u1 <(tar -tzf ~/downloads/pex-2.1.156.tar.gz | sort) <(tar -tzf dist/pex-2.1.156.tar.gz | grep -v -E "/$" | sort)
    --- /dev/fd/63  2024-01-08 01:41:03.688560680 -0800
    +++ /dev/fd/62  2024-01-08 01:41:03.688560680 -0800
    @@ -1,11 +1,6 @@
    -pex-2.1.156/.github/FUNDING.yml
    -pex-2.1.156/.github/workflows/ci.yml
    -pex-2.1.156/.github/workflows/release.yml
    -pex-2.1.156/.gitignore
    -pex-2.1.156/.readthedocs.yaml
     pex-2.1.156/CHANGES.md
     pex-2.1.156/LICENSE
    +pex-2.1.156/MANIFEST.in
     pex-2.1.156/PKG-INFO
     pex-2.1.156/README.rst
    -pex-2.1.156/RELEASE.rst
     pex-2.1.156/docker/base/Dockerfile
    @@ -278,2 +273,3 @@
     pex-2.1.156/pex/vendor/_vendored/pip/.prefix/bin/pip3.8
    +pex-2.1.156/pex/vendor/_vendored/pip/__init__.py
     pex-2.1.156/pex/vendor/_vendored/pip/pip-20.3.4.dist-info/INSTALLER
    @@ -801,2 +797,4 @@
     pex-2.1.156/scripts/typecheck.py
    +pex-2.1.156/setup.cfg
    +pex-2.1.156/setup.py
     pex-2.1.156/testing/__init__.py
    ```
+ wheel:
    ```diff
    $ diff -u1 <(zipinfo -1 ~/downloads/pex-2.1.156-py2.py3-none-any.whl | sort) <(zipinfo -1 dist/pex-2.1.156-py2.py3-none-any.whl | grep -v -E "/$" |
    sort)
    --- /dev/fd/63  2024-01-08 01:42:12.688549984 -0800
    +++ /dev/fd/62  2024-01-08 01:42:12.688549984 -0800
    @@ -5,2 +5,3 @@
     pex-2.1.156.dist-info/entry_points.txt
    +pex-2.1.156.dist-info/top_level.txt
     pex/__init__.py
    @@ -256,2 +257,3 @@
     pex/vendor/_vendored/pip/.prefix/bin/pip3.8
    +pex/vendor/_vendored/pip/__init__.py
     pex/vendor/_vendored/pip/pip-20.3.4.dist-info/INSTALLER
    ```